web/src/engine/websites/Tapread.ts

```diff
@@ -0,0 +1,98 @@
+import { Tags } from '../Tags';
+import icon from './Tapread.webp';
+import { Chapter, DecoratableMangaScraper, Manga, Page, type MangaPlugin } from '../providers/MangaPlugin';
+import * as Common from './decorators/Common';
+import { FetchCSS, FetchJSON, FetchRequest } from './../FetchProvider';
+
+const imgURL = 'https://static.tapread.com';
+
+type APIMangas = {
+    code: number
+    message: string
+    msg: string
+    result: {
+        moreDetailList: {
+            comicId: number,
+            comicName: string
+        }[]
+    }
+};
+
+type APIChapters = {
+    code: number
+    message: string
+    msg: string
+    result: {
+        chapterList: {
+            chapterNo: number,
+            chapterId: number,
+            comicId: number,
+            chapterName: string
+        }[]
+    }
+};
+
+type APIPages = {
+    code: number
+    message: string
+    msg: string
+    result: {
+        imgList: {
+            imgUrl: string,
+            imgWebpUrl: string,
+        }[]
+    }
+};
+
+@Common.ImageAjax()
+
+export default class extends DecoratableMangaScraper {
+
+    public constructor() {
+        super('tapread', `Tapread`, 'https://www.tapread.com', Tags.Language.English, Tags.Source.Official, Tags.Media.Manga, Tags.Media.Manhua, Tags.Media.Manhwa);
+    }
+
+    public override get Icon() {
+        return icon;
+    }
+
+    public override ValidateMangaURL(url: string): boolean {
+        return /^https?:\/\/www\.tapread\.com\/comic\/detail\//.test(url);
+    }
+
+    public override async FetchManga(provider: MangaPlugin, url: string): Promise<Manga> {
+        const id = url.split('/').pop();
+        const data = await FetchCSS<HTMLDivElement>(new FetchRequest(url), 'div.book-container div.book-info div.book-name');
+        return new Manga(this, provider, id, data.pop().textContent.trim());
+    }
+
+    public override async FetchMangas(provider: MangaPlugin): Promise<Manga[]> {
+        const mangalist = [];
+        for (let page = 1, run = true; run; page++) {
+            const mangas = await this.getMangasFromPage(page, provider);
+            mangas.length > 0 ? mangalist.push(...mangas) : run = false;
+        }
+        return mangalist;
+    }
+
+    private async getMangasFromPage(page: number, provider: MangaPlugin): Promise<Manga[]> {
+        const url = new URL('/comic/moredetail?pageNo=' + page, this.URI).href;
+        const request = new FetchRequest(url);
+        const data = await FetchJSON<APIMangas>(request);
```

Comment:
May consider object decomposition `const { code, result: { chapterList }} = await FetchJSON<APIMangas>(request);`

Reply:
Sure but, why and how ?  I am not really familiar with that, sorry.

Reply:
It may improve readability e.g., when (re-)using multiple nested properties:
```javascript
const data = Fetch<T>();
return data.result.error ? [] : data.result.chapters.map(...);
// or
const { result: { error, chapters } } = Fetch<T>();
return error ? [] : chapters.map(...);
```

Reply:
problem : code is always 200, result can be null (ie page 6 of manga request)

I cant do ```  return result? ....``` and anyway, exception occurs at FetchJSON because there is no "moreDetails" in result..
```

    private async getMangasFromPage(page: number, provider: MangaPlugin): Promise<Manga[]> {
        const url = new URL('/comic/moredetail?pageNo=' + page, this.URI).href;
        const request = new FetchRequest(url);
        const { result: { moreDetailList } } = await FetchJSON<APIMangas>(request);
        return moreDetailList ? moreDetailList.map(element => new Manga(this, provider, String(element.comicId), element.comicName.trim())) : [];
    }
```